### PR TITLE
Feature/plottinggl

### PR DESCRIPTION
**Pull request by martinfalk**
various fixes, mostly plotting related.

- made FrameBufferObject, TextRenderer, and AxisRenderer move-assignable 
    * deleted copy constructor and assignment operator
- Default RenderContext activated
    * before adding properties to a ListPropertyWidget
    * before removing properties from a ListPropertyWidget
    * before adding a visualizer
- more fixes (TextRenderer, ScatterPlotGL, ...)

### CODE REVIEW

File: modules/plottinggl/src/utils/axisrenderer.cpp [hunk 1]
```diff
@@ -190,11 +212,11 @@ void AxisRendererBase::renderAxis(Camera* camera, const vec3& start, const vec3&
         drawer.draw();
     };
 
-    drawMesh(axisMesh, settings_.getWidth(), false);
+    drawMesh(axisMesh, settings_.get().getWidth(), false);
```
Comment (ResearchDaniel) [inline modules/plottinggl/src/utils/axisrenderer.cpp:194]: Possible to use settings_->getXX instead of settings.get().getXX

File: modules/fontrendering/include/modules/fontrendering/textrenderer.h [hunk 2]
```diff
@@ -316,7 +316,7 @@ class IVW_MODULE_FONTRENDERING_API TextRenderer {
     int fontSize_;        //<! font size in pixel
     double lineSpacing_;  //!< spacing between two lines in percent (default = 0.2)
 
-    const int glyphMargin_;
+    static constexpr int glyphMargin_ = 2;
```
Comment (ResearchDaniel) [inline modules/fontrendering/include/modules/fontrendering/textrenderer.h]: Is this really the same for all instances? 
That probably deserves a comment in that case.

File: modules/qtwidgets/src/properties/collapsiblegroupboxwidgetqt.cpp [hunk 3]
```diff
@@ -515,6 +516,7 @@ void CollapsibleGroupBoxWidgetQt::addButtonLayout(QGridLayout* layout, int row,
     auto removePropertyBtn = createButton("removeListItemButton", str);
 
     connect(removePropertyBtn, &QToolButton::clicked, this, [layout, prop, buttonWidget]() {
+        RenderContext::getPtr()->activateDefaultRenderContext();
```
Comment (ResearchDaniel) [inline modules/qtwidgets/src/properties/collapsiblegroupboxwidgetqt.cpp:521]: This (workaround) deserves a comment

File: modules/qtwidgets/src/properties/listpropertywidgetqt.cpp [hunk 4]
```diff
@@ -71,6 +72,7 @@ ListPropertyWidgetQt::ListPropertyWidgetQt(ListProperty* property)
 
     connect(addItemButton_, &QToolButton::clicked, this, [&]() {
         if (listProperty_ && listProperty_->getPrefabCount() > 0) {
+            RenderContext::getPtr()->activateDefaultRenderContext();
```
Comment (ResearchDaniel) [inline modules/qtwidgets/src/properties/listpropertywidgetqt.cpp:77]: This (workaround) deserves a comment

**Review by ResearchDaniel — DISMISSED** [review]
Some changes that could be made

File: modules/plottinggl/src/utils/axisrenderer.cpp (hunk 1 above)
Comment (martinfalk) [reply modules/plottinggl/src/utils/axisrenderer.cpp:194]: `settings_`  is a `std::reference_wrapper<AxisSettings>` and thus needs `get()` to get the reference.

File: modules/fontrendering/include/modules/fontrendering/textrenderer.h (hunk 2 above)
Comment (martinfalk) [reply modules/fontrendering/include/modules/fontrendering/textrenderer.h]: it was a `const 2` previously set in the constructor. And the name is exactly what it is. The margin/spacing around a glyph in the atlas to prevent interpolation artifacts.

File: modules/qtwidgets/src/properties/collapsiblegroupboxwidgetqt.cpp (hunk 3 above)
Comment (martinfalk) [reply modules/qtwidgets/src/properties/collapsiblegroupboxwidgetqt.cpp:521]: right. Or rather a better concept of not using (and enforcing) GL-related things within properties :)

File: modules/fontrendering/include/modules/fontrendering/textrenderer.h (hunk 2 above)
Comment (martinfalk) [reply modules/fontrendering/include/modules/fontrendering/textrenderer.h]: fixed.

File: modules/qtwidgets/src/properties/collapsiblegroupboxwidgetqt.cpp (hunk 3 above)
Comment (martinfalk) [reply modules/qtwidgets/src/properties/collapsiblegroupboxwidgetqt.cpp:521]: fixed

File: modules/opengl/src/buffer/framebufferobject.cpp [hunk 5]
```diff
@@ -78,7 +78,7 @@ void FrameBufferObject::defineDrawBuffers() {
 }
 
 void FrameBufferObject::deactivate() {
-    if (isActive()) {
+    if (isActive() && (static_cast<GLuint>(prevFbo_) != id_)) {
```
Comment (petersteneteg) [inline modules/opengl/src/buffer/framebufferobject.cpp]: Should the same thing be done in activate?

File: modules/plotting/src/properties/marginproperty.cpp [hunk 6]
```diff
@@ -42,10 +42,10 @@ MarginProperty::MarginProperty(
     float left, InvalidationLevel invalidationLevel /*= InvalidationLevel::InvalidOutput*/,
     PropertySemantics semantics /*= PropertySemantics::Default*/)
     : CompositeProperty(identifier, displayName)
-    , top_("top", "Top", top, 0.0f, 100.0f, 1.0f, invalidationLevel, semantics)
-    , right_("right_", "Right", right, 0.0f, 100.0f, 1.0f, invalidationLevel, semantics)
-    , bottom_("bottom", "Bottom", bottom, 0.0f, 100.0f, 1.0f, invalidationLevel, semantics)
-    , left_("left", "Left", left, 0.0f, 100.0f, 1.0f, invalidationLevel, semantics) {
+    , top_("top", "Top", top, 0.0f, 200.0f, 1.0f, invalidationLevel, semantics)
```
Comment (petersteneteg) [inline modules/plotting/src/properties/marginproperty.cpp]: set the constraint ConstraintBehavior to ignore on max

File: modules/plottinggl/src/processors/scatterplotmatrixprocessor.cpp [hunk 7]
```diff
@@ -179,7 +180,7 @@ ScatterPlotMatrixProcessor::ScatterPlotMatrixProcessor()
 }
 
 template <typename T>
-struct RangeIterator : public std::iterator<std::forward_iterator_tag, T, T, const T *, T> {
+struct RangeIterator : public std::iterator<std::forward_iterator_tag, T, T, const T*, T> {
```
Comment (petersteneteg) [inline modules/plottinggl/src/processors/scatterplotmatrixprocessor.cpp]: remove unused

File: modules/plottinggl/src/utils/axisrenderer.cpp [hunk 8]
```diff
@@ -122,6 +122,28 @@ AxisRendererBase::AxisRendererBase(const AxisSettings& settings)
 AxisRendererBase::AxisRendererBase(const AxisRendererBase& rhs)
     : settings_{rhs.settings_}, shaders_{getShaders()} {}
 
+AxisRendererBase::AxisRendererBase(AxisRendererBase&& rhs)
```
Comment (petersteneteg) [inline modules/plottinggl/src/utils/axisrenderer.cpp]: use defualt?

File: modules/plottinggl/src/utils/axisrenderer.cpp [hunk 9]
```diff
@@ -122,6 +122,28 @@ AxisRendererBase::AxisRendererBase(const AxisSettings& settings)
...
+    : settings_{rhs.settings_}
+    , textRenderer_{std::move(rhs.textRenderer_)}
+    , quadRenderer_{std::move(rhs.quadRenderer_)}
+    , axisPickingId_{rhs.axisPickingId_}
+    , meshes_{std::move(rhs.meshes_)}
+    , caption_{std::move(rhs.caption_)}
+    , shaders_{std::move(rhs.shaders_)} {}
+
+AxisRendererBase& AxisRendererBase::operator=(AxisRendererBase&& rhs) noexcept {
```
Comment (petersteneteg) [inline modules/plottinggl/src/utils/axisrenderer.cpp]: use default?

File: modules/plottinggl/src/utils/axisrenderer.cpp [hunk 10]
```diff
@@ -209,9 +231,20 @@ AxisRenderer::AxisRenderer(const AxisSettings& settings)
                        [&](auto&& p) { return p.second; });
     }} {}
 
-void AxisRenderer::render(const size2_t& outputDims, const size2_t& startPos, const size2_t& endPos,
+AxisRenderer::AxisRenderer(AxisRenderer&& rhs)
+    : AxisRendererBase(rhs), labels_{std::move(rhs.labels_)} {}
```
Comment (petersteneteg) [inline modules/plottinggl/src/utils/axisrenderer.cpp]: needs AxisRendererBase(std::move(rhs)) or just use default

File: modules/plottinggl/src/utils/axisrenderer.cpp [hunk 11]
```diff
@@ -209,9 +231,20 @@ AxisRenderer::AxisRenderer(const AxisSettings& settings)
...
+
+AxisRenderer& AxisRenderer::operator=(AxisRenderer&& rhs) noexcept {
```
Comment (petersteneteg) [inline modules/plottinggl/src/utils/axisrenderer.cpp]: default

File: modules/plottinggl/src/utils/axisrenderer.cpp [hunk 12]
```diff
@@ -358,9 +393,20 @@ AxisRenderer3D::AxisRenderer3D(const AxisSettings& settings)
                        [](auto& tick) { return tick.second; });
     }} {}
 
+AxisRenderer3D::AxisRenderer3D(AxisRenderer3D&& rhs)
```
Comment (petersteneteg) [inline modules/plottinggl/src/utils/axisrenderer.cpp]: same as above

**Review by petersteneteg — CHANGES_REQUESTED** [review]

File: modules/plottinggl/src/utils/axisrenderer.cpp (hunk 8 above)
Comment (martinfalk) [reply modules/plottinggl/src/utils/axisrenderer.cpp]: with default, the following assertions fail:
```c++
    static_assert(std::is_move_assignable_v<AxisRendererBase>);
    static_assert(std::is_move_assignable_v<AxisRenderer>);
```

File: modules/opengl/src/buffer/framebufferobject.cpp (hunk 5 above)
Comment (martinfalk) [reply modules/opengl/src/buffer/framebufferobject.cpp]: potentially yes. The [OpenGL specs](https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/glBindFramebuffer.xhtml) are not referring to binding the same fbo.

File: modules/plottinggl/src/utils/axisrenderer.cpp (hunk 10 above)
Comment (martinfalk) [reply modules/plottinggl/src/utils/axisrenderer.cpp]: fixed

File: modules/plottinggl/src/utils/axisrenderer.cpp (hunk 12 above)
Comment (martinfalk) [reply modules/plottinggl/src/utils/axisrenderer.cpp]: fixed

**Review by ResearchDaniel — DISMISSED** [review]

File: modules/opengl/src/image/imagecompositor.cpp [hunk 13]
```diff
@@ -69,4 +85,36 @@ void ImageCompositor::composite(ImageInport& source, ImageOutport& target, Image
...
+        colorTex_ = *layer->getTexture();
+    }
+    {
+        auto layer = glrep->getDepthLayerGL();
+        IVW_ASSERT(layer, "no depth layer in target image");
+        depthTex_ = *layer->getTexture();
+    }
+    {
+        auto layer = glrep->getPickingLayerGL();
+        IVW_ASSERT(layer, "no picking layer in target image");
```
Comment (ResearchDaniel) [inline modules/opengl/src/image/imagecompositor.cpp:99]: This will be the case if picking is not enabled in the settings so this should not be an assert.

File: modules/opengl/src/image/imagecompositor.cpp [hunk 14]
```diff
@@ -53,14 +64,19 @@ void ImageCompositor::composite(const Image& source, Image& target, ImageType ty
 
 void ImageCompositor::composite(ImageInport& source, ImageOutport& target, ImageType type) {
     if (source.isReady() && target.hasData()) {
+        copyTextures(*target.getData());
```
Comment (ResearchDaniel) [inline modules/opengl/src/image/imagecompositor.cpp:67]: is it not possible to call composite(*source.getData(), *target.getData(), type) to avoid code duplication?

File: modules/opengl/src/image/imagecompositor.cpp [hunk 15]
```diff
@@ -53,14 +64,19 @@ void ImageCompositor::composite(const Image& source, Image& target, ImageType ty
...
+
         utilgl::activateTarget(target, type);
         shader_.activate();
+        utilgl::GlBoolState depthTest(GL_DEPTH_TEST, false);
+
+        GLint currentFbo = 0;
```
Comment (ResearchDaniel) [inline modules/opengl/src/image/imagecompositor.cpp]: Remove?

File: modules/opengl/include/modules/opengl/image/imagecompositor.h [hunk 16]
```diff
@@ -50,7 +52,13 @@ class IVW_MODULE_OPENGL_API ImageCompositor {
     void composite(ImageInport& source, ImageOutport& target, ImageType type);
 
 private:
+    void copyTextures(const Image& target);
+    void bindTextures(TextureUnitContainer& cont, const std::string& id);
+
     Shader shader_;
+    Texture2D colorTex_;
```
Comment (ResearchDaniel) [inline modules/opengl/include/modules/opengl/image/imagecompositor.h:59]: is not the interface for `composite(...)` incorrect?
It is not obvious that it should combine source and target into target.

It should be `composite(const Image& first, const Image& second, Image& target, ImageType type);` in my opinion.
Not part of this PR, but the documentation about what the supplied fragment shader should take as input and produce is missing.

**Review by ResearchDaniel — CHANGES_REQUESTED** [review]
I think that the ImageCompositor has the wrong interface, so the intermediate copying is the wrong fix

File: modules/opengl/include/modules/opengl/image/imagecompositor.h (hunk 16 above)
Comment (martinfalk) [reply modules/opengl/include/modules/opengl/image/imagecompositor.h:59]: This code was >5 years old and it broke when cleaning up the FBO state. That is two regression tests using the Geometry Picking processor (aka. `meshpicking.h`) were failing.
This is nothing more than a hot fix. I don't even know why we would need such a special compositing in the first place.
Even the shader looks like it does nothing special, so ActivateAndCopy might actually do the expected thing.

https://github.com/inviwo/inviwo/blob/master/modules/opengl/glsl/composite.frag

File: modules/opengl/src/image/imagecompositor.cpp (hunk 15 above)
Comment (martinfalk) [reply modules/opengl/src/image/imagecompositor.cpp]: hmm, left-over from debugging.

File: modules/opengl/src/image/imagecompositor.cpp (hunk 14 above)
Comment (martinfalk) [reply modules/opengl/src/image/imagecompositor.cpp:67]: not really. `target.getData()` returns a `std::shared_ptr<const Image>`, note the `const` here. Hence, it will not be possible to render into the image.

File: modules/opengl/src/image/imagecompositor.cpp (hunk 13 above)
Comment (martinfalk) [reply modules/opengl/src/image/imagecompositor.cpp:99]: Doesn't an image "always" have a picking buffer, even when disabled? If not, then this must be documented somewhere.

Comment (martinfalk) [reply modules/opengl/src/image/imagecompositor.cpp:99]: Having assertions enforce and picking turned off, image outports still have a picking buffer.

Try: Random Sphere Generator, Sphere Renderer, Canvas

Canvas shows the "picking buffer"
image

Comment (martinfalk) [reply modules/opengl/src/image/imagecompositor.cpp:99]: see also info from `Image Information` processor
image

**Comment by martinfalk** [comment]
> I think that the ImageCompositor has the wrong interface, so the intermediate copying is the wrong fix

It has been like this for years ;-) Processors using this ImageCompositor are:
- `DrawLines`
- `DrawPoints`
- `ImageCompositeProcessorGL`
- `MeshPicking`

Potentially all these processors do weird stuff to get the compositing done. For example, why not use regular blending with the depth test?

File: modules/opengl/src/image/imagecompositor.cpp (hunk 13 above)
Comment (ResearchDaniel) [reply modules/opengl/src/image/imagecompositor.cpp:99]: Hmm, then I think that the picking setting can be removed

**Review by ResearchDaniel — DISMISSED** [review]
Ok, let's revise the ImageComposite in another PR

File: modules/plottinggl/src/utils/axisrenderer.cpp (hunk 8 above)
Comment (petersteneteg) [reply modules/plottinggl/src/utils/axisrenderer.cpp]: why? your implementation looks like a default one?

Comment (petersteneteg) [reply modules/plottinggl/src/utils/axisrenderer.cpp]: or have you skipped some members?

File: modules/opengl/src/image/imagecompositor.cpp (hunk 13 above)
Comment (petersteneteg) [reply modules/opengl/src/image/imagecompositor.cpp:99]: the settings has nothing to do with the image layer, but with propagation of picking events

File: modules/opengl/include/modules/opengl/image/imagecompositor.h (hunk 16 above)
Comment (petersteneteg) [reply modules/opengl/include/modules/opengl/image/imagecompositor.h:59]: where is it use can we remove it?

**Comment by petersteneteg** [comment]
> * `DrawLines`
> * `DrawPoints`

There two are very old and also quite embarrassing. I would prefer to remove them

> * `ImageCompositeProcessorGL`
> * `MeshPicking`

These two we could maybe update and fix?

File: modules/plottinggl/src/utils/axisrenderer.cpp (hunk 8 above)
Comment (martinfalk) [reply modules/plottinggl/src/utils/axisrenderer.cpp]: using default now after fixing all members.

**Review by petersteneteg — DISMISSED** [review]

File: modules/opengl/src/buffer/framebufferobject.cpp (hunk 5 above)
Comment (martinfalk) [reply modules/opengl/src/buffer/framebufferobject.cpp]: addressed in 23d2fbecb6fb40619e99ea94dc8b49243c356f16

**Review by petersteneteg — APPROVED** [review]
